src/dodal/devices/oav/snapshots/snapshot_with_grid.py

```diff
@@ -0,0 +1,64 @@
+from io import BytesIO
+from os.path import join as path_join
+
+import aiofiles
+from ophyd_async.core import soft_signal_rw
+from PIL.Image import Image
+
+from dodal.devices.areadetector.plugins.MJPG import MJPG
+from dodal.devices.oav.snapshots.grid_overlay import (
+    add_grid_border_overlay_to_image,
+    add_grid_overlay_to_image,
+)
+from dodal.log import LOGGER
+
+
+class SnapshotWithGrid(MJPG):
+    def __init__(self, prefix: str, name: str = "") -> None:
+        super().__init__(prefix, name)
+
+        self.top_left_x = soft_signal_rw(int)
+        self.top_left_y = soft_signal_rw(int)
+        self.box_width = soft_signal_rw(int)
+        self.num_boxes_x = soft_signal_rw(int)
+        self.num_boxes_y = soft_signal_rw(int)
+
+        self.last_path_outer = soft_signal_rw(str)
+        self.last_path_full_overlay = soft_signal_rw(str)
+
+    async def _save_grid_to_image(self, image: Image, path: str):
+        buffer = BytesIO()
+        image.save(buffer, format="png")
+        async with aiofiles.open(path, "wb") as fh:
+            await fh.write(buffer.getbuffer())
```

Comment:
Should: Can we pull this out into a util function and give it a better name, maybe `asyncio_save_image`? Then we can reuse it in `MJPG` too. Additionally make the format a parameter we can pass in